src/sentry/tasks/relay.py

```diff
@@ -22,72 +22,28 @@ def update_config_cache(
     :param project_id: The project for which to invalidate configs.
     :param generate: If `True`, caches will be eagerly regenerated, not only
         invalidated.
     :param update_reason: A string to set as tag in sentry.
     """
-
-    from sentry.models import Project, ProjectKey, ProjectKeyStatus
-    from sentry.relay import projectconfig_cache
-    from sentry.relay.config import get_project_config
-
-    if project_id:
-        set_current_event_project(project_id)
-
-    if organization_id:
-        # Cannot use bind_organization_context here because we do not have a
-        # model and don't want to fetch one
-        sentry_sdk.set_tag("organization_id", organization_id)
-
-    if public_key:
-        sentry_sdk.set_tag("public_key", public_key)
+    validate_args(organization_id, project_id, public_key)
 
     sentry_sdk.set_tag("update_reason", update_reason)
     sentry_sdk.set_tag("generate", generate)
 
     try:
-        if organization_id:
-            projects = list(Project.objects.filter(organization_id=organization_id))
-            keys = list(ProjectKey.objects.filter(project__in=projects))
-        elif project_id:
-            projects = [Project.objects.get(id=project_id)]
-            keys = list(ProjectKey.objects.filter(project__in=projects))
-        elif public_key:
-            try:
-                keys = [ProjectKey.objects.get(public_key=public_key)]
-            except ProjectKey.DoesNotExist:
-                # In this particular case, where a project key got deleted and
-                # triggered an update, we know that key doesn't exist and we want to
-                # avoid creating more tasks for it.
-                #
-                # In other similar cases, like an org being deleted, we potentially
-                # cannot find any keys anymore, so we don't know which cache keys
-                # to delete.
-                projectconfig_cache.set_many({public_key: {"disabled": True}})
-                return
-        else:
-            assert False
-
-        if generate:
-            config_cache = {}
-            for key in keys:
-                if key.status != ProjectKeyStatus.ACTIVE:
-                    project_config = {"disabled": True}
-                else:
-                    project_config = get_project_config(
-                        key.project, project_keys=[key], full_config=True
-                    ).to_dict()
-                config_cache[key.public_key] = project_config
-
-            projectconfig_cache.set_many(config_cache)
-        else:
-            cache_keys_to_delete = []
-            for key in keys:
-                cache_keys_to_delete.append(key.public_key)
+        keys = project_keys_to_update(
+            organization_id=organization_id, project_id=project_id, public_key=public_key
+        )
 
-            projectconfig_cache.delete_many(cache_keys_to_delete)
+        # TODO: remove this if statement before this PR is merged.
```

Comment:
Is this TODO still relevant? I would keep the `generate` condition for now. If during the deploy there is a new async worker picking up an invalidation task from an old web worker, we still want to respect that flag.